File: src/main/java/aurocosh/divinefavor/common/lib/distributed_random/DistributedRandom.java
package aurocosh.divinefavor.common.lib.distributed_random;

import java.util.HashMap;
import java.util.Map;

public class DistributedRandom {
    private Map<Integer, Double> distributions;
    private double distributionsSum;

    public DistributedRandom() {
        distributions = new HashMap<>();
    }

    public void addNumber(int value, double distribution) {
        if (distributions.get(value) != null)
            distributionsSum -= distributions.get(value);
        distributions.put(value, distribution);
        distributionsSum += distribution;
    }

    public int getRandomNumber() {
        double rand = Math.random();
        double ratio = 1.0f / distributionsSum;
        double tempDist = 0;
        for (Map.Entry<Integer, Double> entry : distributions.entrySet()) {
            tempDist += entry.getValue();
            if (rand / ratio <= tempDist)
                return entry.getKey();
        }
        return 0;
    }

    public void clear() {
        distributions.clear();
        distributionsSum = 0;
    }
}

File: src/main/kotlin/aurocosh/divinefavor/common/lib/distributed_random/DistributedRandom.kt
package aurocosh.divinefavor.common.lib.distributed_random

import java.util.*

class DistributedRandom {
    private val distributions: MutableMap<Int, Double>
    private var distributionsSum: Double = 0.toDouble()

    val randomNumber: Int
        get() {
            val rand = Math.random()
            val ratio = 1.0f / distributionsSum
            var tempDist = 0.0
            for ((key, value) in distributions) {
                tempDist += value
                if (rand / ratio <= tempDist)
                    return key
            }
            return 0
        }

    init {
        distributions = HashMap()
    }

    fun addNumber(index: Int, distribution: Double) {
        distributionsSum -= distributions[index] ?: 0.0
        distributions[index] = distribution
        distributionsSum += distribution
    }

    fun clear() {
        distributions.clear()
        distributionsSum = 0.0
    }
}
